ring/lifecycler_test.go

```diff
@@ -1268,6 +1292,173 @@ func TestJoinInJoiningState(t *testing.T) {
 	})
 }
 
+func TestWaitBeforeJoining(t *testing.T) {
+	ringStore, closer := consul.NewInMemoryClient(GetCodec(), log.NewNopLogger(), nil)
+	t.Cleanup(func() { assert.NoError(t, closer.Close()) })
+
+	var ringConfig Config
+	flagext.DefaultValues(&ringConfig)
+	ringConfig.KVStore.Mock = ringStore
+
+	r, err := New(ringConfig, "ingester", ringKey, log.NewNopLogger(), nil)
+	require.NoError(t, err)
+	require.NoError(t, services.StartAndAwaitRunning(context.Background(), r))
+	defer services.StopAndAwaitTerminated(context.Background(), r) //nolint:errcheck
+
+	err = r.KVClient.CAS(context.Background(), ringKey, func(in interface{}) (interface{}, bool, error) {
+		r := &Desc{
+			Ingesters: map[string]InstanceDesc{
+				instanceName(0, 1): {
+					State:  ACTIVE,
+					Tokens: []uint32{1, 2, 3},
+				},
+				instanceName(1, 1): {
+					State:  ACTIVE,
+					Tokens: []uint32{4, 5, 6},
+				},
+			},
+		}
+
+		return r, true, nil
+	})
+	require.NoError(t, err)
+
+	targetZone := zone(1)
+	spreadMinimizingZones := []string{zone(1), zone(2), zone(3)}
+	canJoinTimeout := 5 * time.Second
+	spreadMinimizingTokenGenerator := func(targetInstanceID string, canJoinEnabled bool) TokenGenerator {
+		tokenGenerator, err := NewSpreadMinimizingTokenGenerator(targetInstanceID, targetZone, spreadMinimizingZones, canJoinEnabled, log.NewNopLogger())
+		require.NoError(t, err)
+		return tokenGenerator
+	}
+
+	tests := map[string]struct {
+		targetInstanceID string
+		tokenGenerator   TokenGenerator
+		timeoutRequired  bool
+		ctx              context.Context
+	}{
+		"RandomTokenGenerator never returns errors": {
```

Comment:
[nit] For all these test cases, the test doesn't check whether returns error or not (it never returns errors), so I think we can improve a bit the description of the test cases. What we're really testing is whether it returns immediately or after the timeout.

Reply:
I forgot to add a test case in which the context gets cancelled, that I have now corrected. 
If you are still not happy with it, I will modify it in a follow-up PR.